src/LuauExt.cpp

```diff
@@ -531,18 +531,41 @@ Luau::ExprOrLocal findExprOrLocalAtPositionClosed(const Luau::SourceModule& sour
 struct FindSymbolReferences : public Luau::AstVisitor
 {
     Luau::Symbol symbol;
-    std::vector<Luau::Location> result{};
+    std::optional<std::vector<lsp::DocumentHighlightKind>> kinds;
+    std::vector<Luau::Location> locations{};
+    bool isModuleImport = false;
+    bool shouldExitCurrentScope = false;
+    size_t symbolDepth = 0;
 
-    explicit FindSymbolReferences(Luau::Symbol symbol)
+    explicit FindSymbolReferences(Luau::Symbol symbol, bool withKinds)
         : symbol(symbol)
+        , kinds(withKinds ? std::optional(std::vector<lsp::DocumentHighlightKind>()) : std::nullopt)
+    {
+    }
+
+    void addResult(Luau::Location location, lsp::DocumentHighlightKind kind)
     {
+        locations.push_back(location);
+        if (kinds)
+            kinds.value().push_back(kind);
     }
 
     bool visit(Luau::AstStatBlock* block) override
     {
+        if (symbol.local && block->location.encloses(symbol.local->location))
+            symbolDepth += 1;
+        auto depth = symbolDepth;
         for (Luau::AstStat* stat : block->body)
         {
             stat->visit(this);
+            if (shouldExitCurrentScope)
+            {
+                shouldExitCurrentScope = false;
+                break;
+            }
+            if (depth != symbolDepth)
+                // This scope is an ancestor of the one in which the local is defined, so we won't find any references here
```

Comment:
This exists for two reasons:
1) it allows us to make the type prefix / module import search sound: we can't keep searching after we've exited the scope in which the module is imported because then, a module imported with the same name could appear, the references to which also would get counted. We're essentially "simulating" type prefixes working like locals with this depth check.
2) it allows us to skip a lot of processing which isn't needed in some cases.

Reply:
I don't think point 2 really applies in this scenario, code execution is pretty fast :) So we don't need to over optimise it until some sort of measurement tells us this is a problem.

TBH, the module imported with the same name seems like such a minor edge case which is why I did not bother fixing it in the first place.

Happy to leave it in for now though. Hopefully that Luau issue gets sorted some time and we don't need to do this.